rust/agama-autoinstall/src/scripts.rs

```diff
@@ -0,0 +1,157 @@
+//
+// This program is free software; you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the Free
+// Software Foundation; either version 2 of the License, or (at your option)
+// any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along
+// with this program; if not, contact SUSE LLC.
+//
+// To contact SUSE LLC about this file by physical or electronic mail, you may
+// find current contact information at www.suse.com.
+
+use std::{
+    fs::{self, create_dir_all, File},
+    io::Write,
+    os::unix::fs::OpenOptionsExt,
+    path::{Path, PathBuf},
+    process::Output,
+};
+
+use agama_lib::utils::Transfer;
+use url::Url;
+
+/// Downloads and runs user-defined scripts.
+pub struct ScriptsRunner {
+    pub path: PathBuf,
+    insecure: bool,
+}
+
+impl ScriptsRunner {
+    /// Creates a new scripts runner.
+    ///
+    /// * path: working directory for the runner.
+    /// * insecure: whether to check certificates when downloading scripts.
+    pub fn new<P: AsRef<Path>>(path: P, insecure: bool) -> Self {
+        Self {
+            path: path.as_ref().to_path_buf(),
+            insecure,
+        }
+    }
+
+    /// Downloads and runs the script from the given URL.
+    ///
+    /// It downloads the script from the given URL to the runner directory.
+    /// It saves the stdout, stderr and exit code to separate files.
+    ///
+    /// * url: script URL.
+    pub fn run(&self, url: &str) -> anyhow::Result<()> {
+        create_dir_all(&self.path)?;
+
+        let file_name = self
+            .file_name_for(&url)
+            .unwrap_or(PathBuf::from("unnamed.sh"));
+
+        let path = self.path.join(&file_name);
+        self.save_script(url, &path)?;
+
+        let output = std::process::Command::new(&path).output()?;
+        self.save_logs(&path, output)?;
+
+        Ok(())
+    }
+
+    pub fn path(&self) -> &PathBuf {
+        &self.path
+    }
+
+    fn file_name_for(&self, url: &str) -> Option<PathBuf> {
+        let parsed = Url::parse(&url).ok()?;
+        let Some(path) = parsed.path_segments() else {
+            return None;
+        };
+
+        path.last().map(|s| PathBuf::from(s))
+    }
+
+    fn save_script(&self, url: &str, path: &PathBuf) -> anyhow::Result<()> {
+        let mut file = Self::create_file(&path, 0o700)?;
+        Transfer::get(url, &mut file, self.insecure)?;
+        Ok(())
+    }
+
+    fn save_logs(&self, path: &Path, output: Output) -> anyhow::Result<()> {
+        if !output.stdout.is_empty() {
+            let mut file = Self::create_file(&path.with_extension("stdout"), 0o600)?;
+            file.write_all(&output.stdout).unwrap();
```

Comment:
```suggestion
            file.write_all(&output.stdout)?;
```

For consistency with the other files.